File: alg_Python/geometry.py
import numpy as np


def fit_line_pca(points):
    """
    Robust PCA line fitting.

    Returns:
        direction (unit vector),
        centroid,
        eigenvalues (sorted descending),
        linearity (lambda0 / lambda1)
    """
    pts = np.asarray(points)

    # Handle irregular input
    if pts.ndim == 1:
        pts = np.array([(p[0], p[1], p[2]) for p in pts])

    if pts.shape[0] < 5:
        return np.zeros(3), np.zeros(3), np.zeros(3), 0.0

    xyz = pts[:, :3]

    # Compute centroid and center data
    centroid = xyz.mean(axis=0)
    centered = xyz - centroid

    # Covariance
    n = centered.shape[0]
    cov = (centered.T @ centered) / (n-1)



    # Eigen decomposition (stable for symmetric matrices)
    eigvals, eigvecs = np.linalg.eigh(cov)

    # Sort descending
    order = np.argsort(eigvals)[::-1]
    eigvals = eigvals[order]
    eigvecs = eigvecs[:, order]

    # Stabilize eigenvalues (prevents collapse during motion/noise)
    eigvals[1] = max(eigvals[1], 1e-3)

    # Principal direction
    direction = eigvecs[:, 0]
    norm = np.linalg.norm(direction)

    if norm < 1e-6:
        return np.zeros(3), centroid, eigvals, 0.0

    direction = direction / norm

    # Enforce consistent direction (reduces angle flipping)
    if direction[0] < 0:
        direction = -direction

    # Linearity metric
    linearity = eigvals[0] / (eigvals[1] + 1e-6)

    return direction, centroid, eigvals, linearity


def compute_angle(direction, reference_axis=None):
    """
    Compute angle (degrees) between barrier direction and reference axis.

    Returns angle in [0, 90].
    """
    if reference_axis is None:
        reference_axis = np.array([0.0, 1.0, 0.0])

    direction = np.asarray(direction)
    reference_axis = np.asarray(reference_axis)

    # Normalize
    dir_norm = np.linalg.norm(direction)
    ref_norm = np.linalg.norm(reference_axis)

    if dir_norm < 1e-6 or ref_norm < 1e-6:
        return 0.0

    direction = direction / dir_norm
    reference_axis = reference_axis / ref_norm

    # Use absolute dot → orientation invariant
    cos_theta = np.clip(abs(direction @ reference_axis), -1.0, 1.0)

    angle = np.degrees(np.arccos(cos_theta))

    return angle

File: alg_C++/src/geometry.cpp
#include "barrier_detection/geometry.hpp"

#include <Eigen/Eigenvalues>
#include <boost/shared_ptr.hpp>
#include <cmath>

// ---------------------------------------------------------------------------
// fit_line_pca
// ---------------------------------------------------------------------------
// Direct port of Python geometry.fit_line_pca().
//
// Key differences from a generic PCA:
//   • Uses the biased-corrected covariance  C = XᵀX / (n-1)  (numpy default)
//   • Clamps λ₁ ≥ 1e-3 to stabilise motion/noise
//   • Enforces direction.x() ≥ 0 for temporal consistency
// ---------------------------------------------------------------------------

PCAResult fit_line_pca(const boost::shared_ptr<pcl::PointCloud<pcl::PointXYZ>> & cloud)
{
    PCAResult result;
    result.direction   = Eigen::Vector3f::Zero();
    result.direction2  = Eigen::Vector3f::Zero();
    result.centroid    = Eigen::Vector3f::Zero();
    result.eigenvalues = Eigen::Vector3f::Zero();
    result.linearity   = 0.0f;

    if (!cloud || static_cast<int>(cloud->size()) < 5) {
        return result;
    }

    const int n = static_cast<int>(cloud->size());

    // ── Centroid ─────────────────────────────────────────────────────────────
    Eigen::Vector3f centroid = Eigen::Vector3f::Zero();
    for (const auto & pt : cloud->points) {
        centroid += Eigen::Vector3f(pt.x, pt.y, pt.z);
    }
    centroid /= static_cast<float>(n);
    result.centroid = centroid;

    // ── Centred data matrix  [n × 3] ─────────────────────────────────────────
    Eigen::MatrixXf X(n, 3);
    for (int i = 0; i < n; ++i) {
        X(i, 0) = cloud->points[i].x - centroid.x();
        X(i, 1) = cloud->points[i].y - centroid.y();
        X(i, 2) = cloud->points[i].z - centroid.z();
    }

    // ── Covariance  C = XᵀX / (n-1) ─────────────────────────────────────────
    Eigen::Matrix3f cov = (X.transpose() * X) / static_cast<float>(n - 1);

    // ── Eigen decomposition (SelfAdjointEigenSolver returns ascending order) ─
    Eigen::SelfAdjointEigenSolver<Eigen::Matrix3f> solver(cov);
    Eigen::Vector3f eigvals = solver.eigenvalues();   // ascending: λ₀ ≤ λ₁ ≤ λ₂
    Eigen::Matrix3f eigvecs = solver.eigenvectors();  // cols correspond to eigvals

    // Reverse to descending (matches np.argsort(eigvals)[::-1])
    Eigen::Vector3f eigvals_desc;
    eigvals_desc << eigvals(2), eigvals(1), eigvals(0);

    Eigen::Matrix3f eigvecs_desc;
    eigvecs_desc.col(0) = eigvecs.col(2);  // principal
    eigvecs_desc.col(1) = eigvecs.col(1);  // second
    eigvecs_desc.col(2) = eigvecs.col(0);  // third

    // ── Stabilise λ₁ ─────────────────────────────────────────────────────────
    if (eigvals_desc(1) < 1e-3f) { eigvals_desc(1) = 1e-3f; }
    result.eigenvalues = eigvals_desc;

    // ── Principal direction ───────────────────────────────────────────────────
    Eigen::Vector3f dir = eigvecs_desc.col(0);
    const float norm = dir.norm();
    if (norm < 1e-6f) { return result; }
    dir /= norm;

    // Enforce consistent sign (reduces angle flipping across frames)
    if (dir.x() < 0.0f) { dir = -dir; }
    result.direction = dir;

    // ── Second direction (for width projection in the node) ───────────────────
    Eigen::Vector3f dir2 = eigvecs_desc.col(1);
    const float norm2 = dir2.norm();
    if (norm2 > 1e-6f) { dir2 /= norm2; }
    result.direction2 = dir2;

    // ── Linearity ─────────────────────────────────────────────────────────────
    result.linearity = eigvals_desc(0) / (eigvals_desc(1) + 1e-6f);

    return result;
}

// ---------------------------------------------------------------------------
// compute_angle
// ---------------------------------------------------------------------------
// Mirrors Python geometry.compute_angle(direction, reference_axis=[0,1,0]).
// Returns angle in [0, 90] degrees (orientation-invariant via |dot|).
// ---------------------------------------------------------------------------

float compute_angle(const Eigen::Vector3f & direction,
                    const Eigen::Vector3f & reference)
{
    const float dir_norm = direction.norm();
    const float ref_norm = reference.norm();

    if (dir_norm < 1e-6f || ref_norm < 1e-6f) { return 0.0f; }

    const Eigen::Vector3f d = direction / dir_norm;
    const Eigen::Vector3f r = reference / ref_norm;

    const float cos_theta = std::clamp(std::abs(d.dot(r)), -1.0f, 1.0f);
    return std::acos(cos_theta) * 180.0f / static_cast<float>(M_PI);
}
